toroid.py

```diff
@@ -0,0 +1,304 @@
+from collections.abc import Callable, Iterable, Mapping
+import math
+import numpy as np
+from numpy import linalg as la
+
+'''A module for modeling Elliptic Toroid surfaces for raytracing-like applications, specifically Celeritas and ORANGE.
+
+This module contains a class "Toroid" which represents an elliptical Toroid surface, with three functions necessary for raytracing:
+ - The first intersection (if any) with the surface along a given ray
+ - The sense of a point relative to the torus, whether inside, on, or outside
+ - The normal vector to the surface at a given point
+
+Note
+----
+Ray-torus intersection and normal formulae taken from Graphics Gems II, Gem V.2 by Joseph M Cychosz, Purdue University
+
+'''
+
+
+class Toroid:
+    '''A class representing a z-axis oriented, origin-centered elliptical toroid in terms of: 
+        r, the major radius (along xy plane); 
+        a, the radius of the rotated ellipse which is parallel to the major radius (xy plane); and 
+        b, the ellipse radius orthogonal to the major radius (z axis). 
+        
+    Class objects also contain three methods for the three surface functions: distance from point
+    along ray, surface sense at point, and surface normal at point.
+
+    Note
+    ----
+    The class also contains a secondary representation of its properties in terms of 3 parameters
+    p, A, and B, for the purpose of solving ray intersection. However, accessing these parameters
+    directly is highly discouraged, as their precise implementation is non-final.
+    
+    '''
+    def __init__(self, r: float, a: float, b: float):
+        '''
+        Parameters
+        ----------
+        r : float
+            The radius from the origin of the toroid to the center of the revolved ellipse, or the "Major Radius", along the xy plane
+        a : float
+            The radius of the revolved ellipse aligned with the major radius, or the "parallel minor radius", along the xy plane
+        b : float
+            The radius of the revolved ellipse orthogonal to the major radius, or the "orthogonal minor radius", aligned with the z-axis
+
+        Attributes
+        ----------
+        TODO: Do you define the effective attributes that the user sees here, or include the private facing ones? I assume the former, in which case it should basically be identical
+        '''
+        self._r = r
+        self._a = a
+        self._b = b
+
+        # From Graphics Gems, form which is more convenient for solving ray intersection
+        self._p = (a*a) / (b*b)
+        self._A = 4*r*r
```

Comment:
You should never have capitol letter variable in Python. I would rename r, a, b as something like "tor_rad", "minor_rad", "major_rad", then use p, a, b for the Graphics Gems variables. Mention that you are storing the variable on the class to avoid recomputing them from successive calls to the ray intersection method.